ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlWebSocketResponseBody.kt

```diff
@@ -41,9 +43,16 @@ internal class CurlWebSocketResponseBody(
 
         val bytesRead = (size * count).toInt()
         val meta = curl_ws_meta(curl)?.pointed ?: error("Missing metadata")
-        val frameData = buffer.readBytes(bytesRead)
 
-        onFrame(frameData, meta.flags)
+        partialFragBuffer += buffer.readBytes(bytesRead)
+
+        // In-case of large payloads being received, libcurl will call the write
+        // callback multiple times with partial fragments that will not have
+        // the CURLWS_CONT bit set, so we buffer internally
+        if (meta.bytesleft == 0L) {
+          onFrame(partialFragBuffer, meta.flags)
+          partialFragBuffer = byteArrayOf()
```

Comment:
is there any failure scenarios where we also need to do this?

Reply:
What do you mean by failure scenarios here? The callback would only be called when we receive data so there's no failure case as such, are you referring to needing this pattern in other places?

Reply:
if the connection drops while receiving data for example, I'd assume we would need to start `partialFragBuffer` from scratch.

Reply:
Ah that makes sense, I'll review the flow in this library and get back to you